Editing TensorFlow tutorial (#1369)
ARMmbed/Handbook: docs/reference/TensorFlow.md

# Machine learning with TensorFlow and Mbed OS
<!--https://github.com/COTASPAR/K66F/blob/master/README.md-->
As an IoT developer, you might think of machine learning as a server-side technology: sensors on your device capture data and send it to the cloud, where Machine Learning (ML) models on hefty machines make sense of it. A network connection is obligatory, and you are going to expect some latency, not to mention hosting costs.

> But with the launch of [TensorFlow Lite for Microcontrollers](https://www.tensorflow.org/lite/microcontrollers), you can run machine learning inference on extremely low-powered devices, like the Cortex-M microcontroller series. You can deploy your ML models to the IoT devices themselves, bring ML closer to your sensors, remove your reliance on a network connection, and skip the round trip to the server for lower latency.

Reviewer: ```suggestion
But with the launch of [TensorFlow Lite for Microcontrollers](https://www.tensorflow.org/lite/microcontrollers), you can run machine learning inference on extremely low-power devices, like Arm Cortex-M series-based microcontrollers. You can deploy your ML models to the IoT devices themselves, bringing ML closer to your sensors, remove your reliance on a network connection, and skip the round trip to the server for lower latency. A primary benefit of doing local inference is that you're reducing your power consumption compared to transferring data over a wireless connection for instance.
```

Reviewer: Also a key benefit of doing local inference is that you're reducing your power consumption compared to transferring data over a wireless connection for instance.

Reviewer 2: The question below on the [VMs](https://github.com/ARMmbed/mbed-os-5-docs/pull/1369#discussion_r512669546) doesn't let me respond so I'll say here: Yes, I  agree. Should we just say tested on Ubuntu 20.04? rather than walk users through how to get Ubuntu on their machine.

* Git.

* xxd.<!--unless is exists by default on all Linux computers?-->

* A development board that can run [TensorFlow Lite for Microcontrollers](https://www.tensorflow.org/lite/microcontrollers#supported_platforms). We tested on the [NXP FRDM-K66F](https://os.mbed.com/platforms/FRDM-K66F/) development board.

* A mini-USB cable.

<!--MarkDown doesn't support embedded YouTube players, which is one reason I removed the videos. Another is that it's better practice to link to official sites, not to videos uploaded by "unofficial" users who may remove them without warning and may, for all we know, be in legal disputes with the companies who created those products (I admit that isn't very likely with "how to install" videos). Also, not sure these videos will work in all countries.-->
### Download, build and install the application

1. Navigate to the directory where you keep code projects.
1. Download the TensorFlow Lite source code:

    ```
    git clone https://github.com/tensorflow/tensorflow.git
    ```

1. Navigate into the project directory:

    ```
    cd tensorflow/tensorflow/lite/micro/mbed/
    ```
    

> 1. Change the DebugLog function to use UnbufferedSerial instead since the Serial API was deprecated in Mbed OS 6:

Reviewer: Can we upstream this? Or link an open PR in the doc?

Reviewer 2: Many of the other TFLu examples still use older versions of Mbed OS and thus if we upstream it that could break those examples. Ideally, all of the examples would be ported to the latest version of mbed. However, I think for now this should be okay while perhaps we try to engage with the TFLu team to port the other examples.

Reviewer 3: As soon as this is published we will start to work with the TF team to discuss next steps.

    ```
1. Build the project

    ```
    cd ../../../../

    make -f tensorflow/lite/micro/tools/make/Makefile TARGET=mbed TAGS="nxp_k66f" generate_micro_speech_mbed_project
    ```
    <!--I got a "no such file or directory error". I think it's because there was already `cd tensorflow` and then the path in the command also has `tensorflow`. Removing `tensorflow` from the path removed the error-->

    This creates a folder in `tensorflow/lite/micro/tools/make/gen/mbed_cortex-m4/prj/micro_speech/mbed` that contains the source and header files for the example application, Mbed OS driver files, and a README.

1. Navigate into the `mbed` folder:

    ```
    cd tensorflow/lite/micro/tools/make/gen/mbed_cortex-m4/prj/micro_speech/mbed
    ```
1. Connect your board to your computer over USB. On your board, when the Ethernet port is facing you, the micro USB port is to its left.
    Your board appears as storage on your computer. If your system does not recognize the board with the `mbed detect` command, follow the instructions for setting up [DAPLink](https://armmbed.github.io/DAPLink/?board=FRDM-K66F).
    
1. Execute the following to flash the application onto the board:

    ```

>     mbed config root .

Reviewer: Probably a shortcoming of the TFLiteU Makefile above, ideally this wouldn't be needed so candidate for upstreaming too.

    mbed deploy
    

>     mbed update mbed-os-6.3.0

Reviewer: 6.4 was just released ;)

Reviewer: Should also be upstreamed

Reviewer 3: @donatieng in order to feature proof this is there a way to update to the latest Mbed version? Perhaps the latest known version without breaking changes.

Reviewer: Unfortunately we don't have a "update to the latest Mbed OS 6 release" command; good feedback though, it would be good to look at that in the new tools

    mbed compile -m K66F -t GCC_ARM --flash --sterm
    ```

    <!--The install page for Mbed CLI now requires 3.7. Why are we asking for 2.7?-->
    <!--need to explain what `config root .` and `deploy` do and why we need them - neither one is part of a standard workflow where you use Mbed CLI to import an application, so this is a special case-->
    <!--and why are we compiling here? We compile again two steps down, with the flash parameter-->

1. Speak into the board's microphone: Saying "Yes" will print "Yes" and "No" will print "No" on the serial port:

    ```

>     Heard yes (208) @116448ms  

Reviewer: What's 208?

Reviewer 2: Below the output (on line 116) we could add:
> the output shows found_command, score and current_time as described in the command_responder.cc file.

To solve this problem, our code runs inference as often as it can<!--here you lost me, probably because I don't deal with ML - is this in the training again? where is it getting new samples?-->, depending on the speed of the device, and averages all of the results within a rolling 1000ms window. The code in the file [recognize_commands.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc) performs this action. When the average for a given category in a set of predictions goes above the threshold<!--lost me here too - are we talking about training or recognising actual users?-->, as defined in [recognize_commands.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/examples/micro_speech/recognize_commands.h), we can assume a valid result.

### Interpreting the results

> The RespondToCommand method in [command_responder.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/examples/micro_speech/command_responder.cc) is called when a command a user speals has been recognised. Currently, this results in a line being printed to the serial port. Later in this guide, we will modify the code to display the result on the screen.

Reviewer: The K66F board doesn't have a screen - do we need a shield?

Reviewer 2: No, it doesn't have a screen nor did we use a shield. This is a mistake, the change to command_responder.cc is for boards with LCD displays. It does not apply to the K66F.

Reviewer: Should that be removed then?

Reviewer 2: Yes

<span class="tips">**Tip:** Building TensorFlow and training the model will each take a couple of hours on an average computer. We will not perform this at this stage. If you are interested in the full flow, including the training of the model, refer to the [Supplementary information: model training](https://developer.arm.com/solutions/machine-learning-on-arm/developer-material/how-to-guides/build-arm-cortex-m-voice-assistant-with-google-tensorflow-lite/supplementary-information-model-training) guide.</span>

## Background: Converting the trained model

To get a converted model that can run on the controller itself from the trained model, we need to run a conversion script: the [TensorFlow Lite converter](https://www.tensorflow.org/lite/convert). This tool makes our model as small and efficient as possible, and converts it to a TensorFlow Lite FlatBuffer. To reduce the size of the model, we used a technique called [quantization](https://www.tensorflow.org/lite/performance/post_training_quantization). All weights and activations in the model get converted from 32-bit floating point format to an 8-bit and fixed-point format, as you can see in the following command:

> <span class="images">![Convert the model to the TensorFlow Lite format code](https://raw.githubusercontent.com/COTASPAR/K66F/master/images/convert_model.png)</span>

Reviewer: Can this image be made part of this PR's scope?

Reviewer 2: This should be hosted on Arm's GitHub and link from there.

Reviewer: Needs to be updated to a relative link and image part of the PR

This conversion will not only reduce the size of the network traffic, but will also avoid the more computationally expensive floating points.

To save time, in this guide we skip the conversion step and instead download a converted model.

## Downloading the model and creating a C file

1. Download the [tiny_conv.tflite file](https://developer.arm.com/-/media/Files/downloads/Machine%20learning%20how-to%20guides/tiny_conv.tflite?revision=495eb362-4325-49b8-b3ba-3141df0c9b95&la=en&hash=0F37BA2C5DE95A1561979CDD18973171767A47C3).
1. Convert this model into a C file that works with Mbed OS projects.

    Use a tool called xxd to generate a file called `model.cc`:

>     ```

Reviewer: What is it? What does it do?

Reviewer 2: Already described above "Convert this model into a C file that works with Mbed OS projects.

Use a tool called xxd to generate a file called model.cc:"

Reviewer: I meant, what is `xxd`? As a user, I don't want to run some arbitrary shell command on my machine without understanding what it does 😉

Reviewer 2: Ah, I see. We could add "[xxd](https://www.tutorialspoint.com/unix_commands/xxd.htm) is a UNIX command-line utility that creates a hex dump of a given file or standard input. It can also convert a hex dump back to its original binary form."

    xxd -i  tiny_conv.tflite > ../micro_features/model.cc
    ```

1. Update `model.cc` so that it is compatible with our code:

    1. Open the file. The top two lines should look similar to the following code, although the exact variable name and hex values may be different:

        ```
        const  unsigned  char  g_model[] DATA_ALIGN_ATTRIBUTE = {  
        0x18, 0x00, 0x00, 0x00, 0x54, 0x46, 0x4c, 0x33, 0x00, 0x00, 0x0e, 0x00,
        ```

    1. Add the `include` from the following snippet and change the variable declaration without changing the hex values:

        ```
        #include "tensorflow/lite/micro/examples/micro_speech/micro_features/model.h"  
        const unsigned char g_tiny_conv_micro_features_model_data[] = {  
        0x18, 0x00, 0x00, 0x00, 0x54, 0x46, 0x4c, 0x33, 0x00, 0x00, 0x0e, 0x00,
        ```

    1. Go to the bottom of the file and find the unsigned int variable:

        ```
        unsigned int tiny_conv_tflite_len = 18216;
        ```

>     1. Change the declaration to the following code, but do not change the number assigned to it, even if your number is different from the one in this guide:

Reviewer: Why? What is it?

Reviewer: "Change the variable's name and constness to"?

Reviewer 2: That number is auto-generated by the converter script and the constant is the length of the model parameters. I suggest the following change "Change the declaration to the following code, but do not change the number assigned to it, as it is automatically generated to indicate the size of the model. Even if your number is different from the one in this guide:"

Reviewer: The "value"? (as opposed to "number")

Reviewer 3: sounds good let's go with value 👍

    You see the following category labels:

    ```
    const char* kCategoryLabels[kCategoryCount] = {  
    "silence",  
    "unknown",  
    "yes",  
    "no",  
    };
    ```

    The code uses this array to map the output of the model to the correct value. Because we specified our wanted_words<!--is that underscore correct?--> as “up, down” in the training script, we should update this array to reflect these words in the same order.

1. Edit the code so that "up" replaces "yes" and "down" replaces "no":

    ```
    const char* kCategoryLabels[kCategoryCount] = {  
    "silence",  
    "unknown",  
    "up",  
    "down",  
    };
    ```

1. Update the code in `command_responder.cc` to reflect these new labels, modifying the if statements and the DisplayStringAt call:

    ```
    void RespondToCommand(tflite::ErrorReporter* error_reporter,  
    int32_t current_time, const char* found_command,  
    uint8_t score, bool is_new_command) {  
        if (is_new_command) {  
            error_reporter->Report("Heard %s (%d) @%dms", found_command, score,  
            current_time);  
        if(strcmp(found_command, "up") == 0) {  

>             lcd. Clear(0xFF0F9D58);  

Reviewer: Where did that LCD screen come from 😄?

Reviewer 2: This is a mistake. This board does not have an LCD screen and the command_responder.cc should not be changed if the model is re-trained.

            lcd.DisplayStringAt(0, LINE(5), (uint8_t *)"Heard up", CENTER_MODE);  
        } else if(strcmp(found_command, "down") == 0) {  
            lcd.Clear(0xFFDB4437);  
            lcd.DisplayStringAt(0, LINE(5), (uint8_t *)"Heard down", CENTER_MODE);  
        } else if(strcmp(found_command, "unknown") == 0) {  
            lcd.Clear(0xFFF4B400);  
            lcd.DisplayStringAt(0, LINE(5), (uint8_t *)"Heard unknown", CENTER_MODE);  
        } else {  
            lcd.Clear(0xFF4285F4);  
            lcd.DisplayStringAt(0, LINE(5), (uint8_t *)"Heard silence", CENTER_MODE);  
            }  
        }  
    }
    ```
<!--I added indents - I assumed their lack was a result of copy/paste destroying something, not a decision-->

1. Navigate back to the mbed directory:

    ```
    cd <path_to_tensorflow>/tensorflow/lite/micro/tools/make/gen/mbed_cortex-m4/prj/micro_speech/mbed
    ```

1. Rebuilt the project:

    ```

>     mbed compile -m K66F -t GCC_ARM

Author: ```suggestion
    mbed compile -m K66F -t GCC_ARM --flash
```

    ```

1. Copy the binary to the USB storage of the device<!--why aren't we using flash?-->, using the same method that you used earlier.

You can now say “up” and “down” to update the display.

## Troubleshooting

We have found some common errors that users face and have listed them here to help you get started with your application as quickly as possible.

- If you encounter:

    ```
    Mbed CLI issues or Error: collect2: error: ld returned 1 exit status
    ```

    Purge the cache with the following command:

    ```
    mbed cache purge
    ```

    You probably also have a stale BUILD folder. Clean up your directory and try again:

    ```
    rm -rf BUILD
    ```

- If your terminal is wrapping your text as show here:

    <span class="images">![Error prompt wrapping around line](https://raw.githubusercontent.com/COTASPAR/K66F/master/images/troubleshooting.png)</span>

    In your terminal, type:

    ```
    export PS1='\u@\h: '
    ```

    For a more minimalist type:

    ```
    export PS1='> '
    ```

- Error: `Requires make version 3.82 or later (current is 3.81)`

>     If you encounter this error, install Brew and Make:

Reviewer: Assuming you're on a Mac

Reviewer 3: In the interest of getting this published, we could remove the below:

> Error: Requires make version 3.82 or later (current is 3.81)
> 
> If you encounter this error, install Brew and Make:
> 
> ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
> 
> brew install make

Reviewer: (or you could say "if you're using macOS")

Reviewer 3: we could, but not sure this is extremely relevant as this was an error that used to occur a while back, it might be obsolete by now

Reviewer: OK, worth removing altogether then!

This is especially exciting for IoT, because lower network use means lower power consumption. You can also better guarantee the security and privacy of your users, since you do not need to send data back to the cloud unless you know for sure that it is relevant.

In this guide, you will learn how to perform machine learning inference on an Arm Cortex-M microcontroller with TensorFlow Lite for Microcontrollers. You will deploy a sample application we wrote that uses the microphone on the K66F and a TensorFlow machine learning model to detect the words “yes” and “no”.

The guide has step by step instructions for installing and running the application, followed by information about how TensorFlow and the application work. Finally, it shows you how to retrain the ML model with new words.

## Using the example application

<!--this needs to move-->
**Important:** We recommend running the following commands from inside the Mbed CLI terminal that gets launched with the Mbed CLI Application. This is because it will be much quicker to set up, because it resolves all your environment dependencies automatically.

### Prerequisites

Here is what you will need to complete the guide:

> * Ubuntu 20.04 LTS. If you are using a Windows machine, install [VirtualBox](https://www.virtualbox.org/wiki/Downloads)<!--originally you said 6.0 - that version's support ended in July. Can they install the latest version, or do they need 6.0 specifically?--> and then [Install Ubunto on it](https://www.virtualbox.org/wiki/Linux_Downloads).

Reviewer 2: Regarding VirtualBox, it should not make a difference as long as they are able to run Ubuntu 20.04 on it. Although their website states "Please also use version 6.0 if you need to run VMs with software virtualization, as this has been discontinued in 6.1. Version 6.0 will remain supported until July 2020". So, I would suggest that for the sake of consistency users stick with version 6.0 to ensure the demo works accordingly. 

 Also, there is a tiny spelling mistake, it should be "[Install Ubuntu on it]", not "[Install Ubunto on it]".

Author: ```suggestion
* Ubuntu 20.04 LTS. If you are using a Windows machine, install [VirtualBox](https://www.virtualbox.org/wiki/Downloads) and then [Install Ubuntu on it](https://www.virtualbox.org/wiki/Linux_Downloads).
```

Reviewer: This all feels very specific. Can we reword and say that this tutorial has been tested on Ubuntu 20.04 - because presumably it will work on other platforms (or even other versions of Ubuntu)? Also VirtualBox is one suggested option but other virtualization SW would work too.

Reviewer 3: Agreed can we say: 

> This tutorial has been tested on Ubuntu 20.04 LTS. If you are using a Windows machine, we recommend using virtualization software to create a virtual Linux machine to run this tutorial.

* [Mbed CLI (version 1.10.4)](../build-tools/install-and-set-up.html).

* The [GNU Arm Embedded](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads) toolchain [that Mbed CLI supports](../build-tools/index.html).<!--Mbed OS supports ARMC 6.14, and many users rely on it. Can they use it?--> For Windows users, if you have already set up the Linux virtual environment, install the toolchain there.

    Configure Mbed CLI to find your GNU installation:

    ```    
    mbed config -G GCC_ARM_PATH <path_to_your_arm_toolchain>/bin
    ```

* Make 3.82 or newer (if you are using an older version, see the troubleshooting section for instructions).

    <span class="notes">**Note**: On a Mac, you might have to use gmake instead of make.</span>

> * Python 2.7. We recommend [using pyenv to manage Python versions](https://pypi.org/project/pyenv/).<!--The install page for Mbed CLI now requires 3.7. Why are we asking for 2.7?-->

Reviewer 2: We tried with Python 3.7 but had issues from the TensorFlow side. See point 5 [here](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/micro/examples/micro_speech#deploy-to-nxp-frdm-k66f).

Author: And that doesn't cause problems with Mbed CLI?

Reviewer: Python 3.x please! What are the issues on the TF-side (apart from outdated tutorial)?

Reviewer 2: Yes, just tested again with a Python 3.7 environment and it worked fine.

* Git.

> * xxd.<!--unless is exists by default on all Linux computers?-->

Reviewer 2: Well, you are right, it does come with Ubuntu although it could be the case some minimalist distros don't include it for some reason. But since the tutorial focuses on Ubuntu we could leave this out.

Author: ```suggestion

```

* A development board that can run [TensorFlow Lite for Microcontrollers](https://www.tensorflow.org/lite/microcontrollers#supported_platforms). We tested on the [NXP FRDM-K66F](https://os.mbed.com/platforms/FRDM-K66F/) development board.

* A mini-USB cable.

> <!--MarkDown doesn't support embedded YouTube players, which is one reason I removed the videos. Another is that it's better practice to link to official sites, not to videos uploaded by "unofficial" users who may remove them without warning and may, for all we know, be in legal disputes with the companies who created those products (I admit that isn't very likely with "how to install" videos). Also, not sure these videos will work in all countries.-->

Reviewer 2: Understood, I'll keep that in mind for future guides.

### Download, build and install the application

1. Navigate to the directory where you keep code projects.
1. Download the TensorFlow Lite source code:

    ```
    git clone https://github.com/tensorflow/tensorflow.git
    ```

1. Navigate into the project directory and build it:

    ```
    cd tensorflow

    make -f tensorflow/lite/micro/tools/make/Makefile TARGET=mbed TAGS="nxp_k66f" generate_micro_speech_mbed_project
    ```

>     <!--I got a "no such file or directory error". I think it's because there was already `cd tensorflow` and then the path in the command also has `tensorflow`. Removing `tensorflow` from the path removed the error-->

Reviewer 2: Okay, just to confirm, the full directory is tensorflow/tensorflow/lite/micro/tools/make/Makefile meaning, you

> git clone https://github.com/tensorflow/tensorflow.git
> cd tensorflow
> make -f tensorflow/lite/micro/tools/make/Makefile TARGET=mbed TAGS="nxp_k66f" generate_micro_speech_mbed_project 

and this should not generate an error. Correct?

Author: When I CDed into tensorflow and then also used it in the path, I think it went tensorflow/tensorflow/lite
It worked only when I removed the tensorflow from the path

Author: I can't tell you if it actually *works* because Brew won't update my Make. But it stopped giving the path error.

Reviewer 2: Ok

    This creates a folder in `tensorflow/lite/micro/tools/make/gen/mbed_cortex-m4/prj/micro_speech/mbed` that contains the source and header files for the example application, Mbed OS driver files, and a README.

1. Navigate into the `mbed` folder:

    ```
    cd tensorflow/lite/micro/tools/make/gen/mbed_cortex-m4/prj/micro_speech/mbed
    ```

1. Execute the following (on an environment that runs Python 2.7):

    ```
    mbed config root .

    mbed deploy

    mbed compile -m K66F -t GCC_ARM
    ```

    <!--The install page for Mbed CLI now requires 3.7. Why are we asking for 2.7?-->

>     <!--need to explain what `config root .` and `deploy` do and why we need them - neither one is part of a standard workflow where you use Mbed CLI to import an application, so this is a special case-->

Reviewer 2: From what I understand, `mbed config root .` sets the current mbed-os folder as the root (meaning `mbed` command actually knows what Mbed OS project we want to run the following commands on) which then allows us to run `mbed deploy` which can pull the libraries needed (in our case in tensorflow/tensorflow/lite/micro/tools/make/gen/mbed_cortex-m4/prj/micro_speech/mbed/mbed-os/requirements.txt) and installs them. I was not aware that this is not a standard workflow, this is the workflow presented in the [original guide hosted on developer.arm.com](https://developer.arm.com/solutions/machine-learning-on-arm/developer-material/how-to-guides/build-arm-cortex-m-voice-assistant-with-google-tensorflow-lite/download-and-build-the-sample-application). If you know of a better way of doing this I'd be glad to explore that workflow and update the guide.

Author: @donatieng can you confirm/deny?

Reviewer: See above, I think this is a shortcoming of the Makefile here

Reviewer 3: are we good here?

>     <!--and why are we compiling here? We compile again two steps down, with the flash parameter-->

Reviewer 2: You are right. This is the format presented in the original guide referenced above. However, we could take that out. One reason could be that we compile to ensure everything is working and only then we compile + flash the device.

Author: The first compile step now has to come out, since it compiled with an old version of Mbed OS.

>     For some compilers<!--we only support two, and you specifically asked for GNU, so in what case will this happen?-->, you may get a compilation error in `mbed_rtc_time.cpp`. Go to  `mbed-os/platform/mbed_rtc_time.h`  and comment out line 32 and line 37:

Reviewer 2: That was coming from the TensorFlow instructions at some point while we were developing the demo. However, I can't seem to find this warning anymore on their repo other than in the K66F instructions which I pushed myself but I just tried this workflow again and it works without commenting anything out so they must have fixed this issue. We can omit this step.

    ```
    //#if !defined(__GNUC__) || defined(__CC_ARM) || defined(__clang__) struct timeval {
    time_t  tv_sec; int32_t tv_usec;
    };
    //#endif
    ```

1. Connect your board to your computer over USB. On your board, when the Ethernet port is facing you, the micro USB port is to its left.

    Your board appears as storage on your computer. If your system does not recognize the board with the `mbed detect` command, follow the instructions for setting up [DAPLink](https://armmbed.github.io/DAPLink/?board=FRDM-K66F).

1. Flash the application to the board:

>     <!--I compiled two steps ago - that probably needs to be removed-->

Reviewer 2: Yes, we could either remove the previous compile step or this one to be less redundant. I'd say it might be better to keep this one as it flashes automatically to the board.

    ```
    mbed compile -m K66F -t GCC_ARM –flash
    ```

> 1. Deploy the example to your K66F<!--but we flshed already. What are we doing here?-->

Reviewer 2: Again, following the format in the original [guide on developer.arm.com](https://developer.arm.com/solutions/machine-learning-on-arm/developer-material/how-to-guides/build-arm-cortex-m-voice-assistant-with-google-tensorflow-lite/single-page). They had two ways of doing things. One where we compile and flash on the same command `mbed compile -m K66F -t GCC_ARM –flash` and one where we just compile then copy the binary over to the board to flash it, as shown below with `cp ./BUILD/K66F/GCC_ARM/mbed.bin /Volumes/K66F/` .

    Copy the binary file that we built earlier to the USB storage.

>     Note: if you have skipped the previous steps<!--which ones?-->, download the [binary file]() to proceed.

Reviewer 2: all the steps up to compiling steps

    The file is at:

    ```
    cp ./BUILD/K66F/GCC_ARM/mbed.bin /Volumes/K66F/
    ```

>     Depending on your operating system <!--aren't we making everyone use Linux?-->, the exact copy command and paths may vary.

Reviewer 2: Good catch. I should have removed this from the original guide.

You can see how the 30ms sample window is moved forward by 20ms each time until it has covered the full one-second sample. The resulting spectrogram is passed into the convolutional model.

### Recognition and windowing

The process of capturing one second of audio and converting it into a spectrogram leaves us with something that our ML model can interpret. The model outputs a probability score for each category it understands (yes, no, unknown, and silence). The probability score indicates whether the audio is likely to belong to that category.

The model was trained on the one-second samples of audio we saw above. In the training data, the word “yes” or “no” is spoken at the start of the sample, and the entire word is contained within that one-second. However, when the application is running, there is no guarantee that a user will begin speaking at the very beginning of our one-second sample. If the user starts saying “yes” at the end of the sample instead of the beginning, the model might not be able to understand the word. This is because the model uses the position of the features within the sample to help predict which word was spoken.

> To solve this problem, our code runs inference as often as it can<!--here you lost me, probably because I don't deal with ML - is this in the training again? where is it getting new samples?-->, depending on the speed of the device, and averages all of the results within a rolling 1000ms window. The code in the file [recognize_commands.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc) performs this action. When the average for a given category in a set of predictions goes above the threshold<!--lost me here too - are we talking about training or recognising actual users?-->, as defined in [recognize_commands.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/examples/micro_speech/recognize_commands.h), we can assume a valid result.

Reviewer 2: This part doesn't refer to training data but rather samples when the model is running on the device. Because we don't know at what point in the "window" a person will say the keyword this means we need to run inference many times so that we effectively take many snapshots of many windows then we analyse those "windows" and then take the average result as that's what is most likely being said. For a deeper explanation you can search "Windowing" on this [book on TinyML](https://tinymlbook.files.wordpress.com/2020/01/tflite_micro_preview.pdf).

Author: ```suggestion
To solve this problem, our code runs inference as often as it can, depending on the speed of the device, and averages all of the results within a rolling 1000ms window. The code in the file [recognize_commands.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/examples/micro_speech/recognize_commands.cc) performs this action. When the average for a given category in a set of predictions goes above the threshold, as defined in [recognize_commands.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/examples/micro_speech/recognize_commands.h), we can assume a valid result.
```

1. Update `model.cc` so that it is compatible with our code:

    1. Open the file. The top two lines should look similar to the following code, although the exact variable name and hex values may be different:

        ```
        const  unsigned  char  g_model[] DATA_ALIGN_ATTRIBUTE = {  
        0x18, 0x00, 0x00, 0x00, 0x54, 0x46, 0x4c, 0x33, 0x00, 0x00, 0x0e, 0x00,
        ```

    1. Add the `include` from the following snippet and change the variable declaration without changing the hex values:

        ```
        #include "tensorflow/lite/micro/examples/micro_speech/micro_features/model.h"  
        const unsigned char g_tiny_conv_micro_features_model_data[] = {  
        0x18, 0x00, 0x00, 0x00, 0x54, 0x46, 0x4c, 0x33, 0x00, 0x00, 0x0e, 0x00,
        ```

    1. Go to the bottom of the file and find the unsigned int variable:

        ```
        unsigned int tiny_conv_tflite_len = 18216;
        ```

    1. Change the declaration to the following code, but do not change the number assigned to it, even if your number is different from the one in this guide:

        ```
        const int g_tiny_conv_micro_features_model_data_len = 18216;
        ```

    1. Save the file.

> 1. Copy the `tiny_conv_micro_features_model_data.cc` <!--where is this file? it's not the one I was just editing-->file into the `tensorflow/tensorflow/lite/micro/tools/make/gen/mbed_cortex-m4/prj/micro_speech/mbed/tensorflow/lite/micro/examples/micro_speech/micro_features` directory.

Reviewer 2: Apologies. This file was originally called `tiny_conv_micro_features_model_data.cc` when the project was[ first made](https://github.com/uTensor/tf_microspeech/blob/master/micro_features/tiny_conv_micro_features_model_data.cc) but this is now just[ `micro_features/model.cc`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/examples/micro_speech/micro_features/model.cc) .

Author: ```suggestion
1. Copy the `model.cc` file into the `tensorflow/tensorflow/lite/micro/tools/make/gen/mbed_cortex-m4/prj/micro_speech/mbed/tensorflow/lite/micro/examples/micro_speech/micro_features` directory.
```

## Modify the application code

If you build and run your code now, your ML model knows the words “up” and “down”, but your application assumes that the words are “yes” and “no”. Let’s update the references and the user interface so that the appropriate words are printed.

1. Navigate to the `tensorflow/lite/micro/examples/micro_speech/` directory and open the file `micro_features/micro_model_settings.cc`.

    You see the following category labels:

    ```
    const char* kCategoryLabels[kCategoryCount] = {  
    "silence",  
    "unknown",  
    "yes",  
    "no",  
    };
    ```

>     The code uses this array to map the output of the model to the correct value. Because we specified our wanted_words<!--is that underscore correct?--> as “up, down” in the training script, we should update this array to reflect these words in the same order.

Reviewer 2: Yes, the underscore is correct but it should actually be all caps "WANTED_WORDS" as on the [python training script](https://github.com/tensorflow/tensorflow/blob/5d4a29eaf590b4a3068ef4d0b7bea9d4f7bd9369/tensorflow/lite/micro/examples/micro_speech/train/train_micro_speech_model.ipynb).

Author: ```suggestion
    The code uses this array to map the output of the model to the correct value. Because we specified our WANTED_WORDS as “up, down” in the training script, we should update this array to reflect these words in the same order.
```

            error_reporter->Report("Heard %s (%d) @%dms", found_command, score,  
            current_time);  
        if(strcmp(found_command, "up") == 0) {  
            lcd. Clear(0xFF0F9D58);  
            lcd.DisplayStringAt(0, LINE(5), (uint8_t *)"Heard up", CENTER_MODE);  
        } else if(strcmp(found_command, "down") == 0) {  
            lcd.Clear(0xFFDB4437);  
            lcd.DisplayStringAt(0, LINE(5), (uint8_t *)"Heard down", CENTER_MODE);  
        } else if(strcmp(found_command, "unknown") == 0) {  
            lcd.Clear(0xFFF4B400);  
            lcd.DisplayStringAt(0, LINE(5), (uint8_t *)"Heard unknown", CENTER_MODE);  
        } else {  
            lcd.Clear(0xFF4285F4);  
            lcd.DisplayStringAt(0, LINE(5), (uint8_t *)"Heard silence", CENTER_MODE);  
            }  
        }  
    }
    ```
<!--I added indents - I assumed their lack was a result of copy/paste destroying something, not a decision-->

1. Navigate back to the mbed directory:

    ```
    cd <path_to_tensorflow>/tensorflow/lite/micro/tools/make/gen/mbed_cortex-m4/prj/micro_speech/mbed
    ```

1. Rebuilt the project:

    ```
    mbed compile -m K66F -t GCC_ARM
    ```

> 1. Copy the binary to the USB storage of the device<!--why aren't we using flash?-->, using the same method that you used earlier.

Reviewer 2: I guess we can leave this out and show it as before. I'm not sure what the logic behind showing two ways of doing it was when the original guide was made.

Author: ```suggestion

```
